### DISCUSSION TITLE
How to save `TaggedFile` to `Vec<u8>` ?

### DISCUSSION THREAD

**Discussion by jerry7536** [Serial-ATA/lofty-rs#293] [Q&A] [answered]
I am new to Rust and want to build a wasm version of this awesome project, see #289 

I am facing a problem of how to save `TaggedFile` to `Vec<u8>`, so that I can send the modified file buffer from wasm to js.

Any help is highly appreciated!

**Comment by Serial-ATA** [comment]
It's a WIP, see #290.

I looked into WASM after you submitted #289. I had no experience with it prior and did not know it doesn't have access to `std::io`.

**Reply by jerry7536** [reply]
Thanks for explaination!

**Comment by jerry7536** [comment]
Hey👋
I upgrade `lofty` to 0.19 and finally write a demo for wasm, but I encountered a question and all the GPT/coplite cannot help, so I'm here asking for help 🙏.

Is there any way to save without cloning the original buffer?

```rs
#[wasm_bindgen]
pub struct MetaFile {
    buffer: Vec<u8>,
    tag: Tag,
}

#[wasm_bindgen]
impl MetaFile {
    #[wasm_bindgen(constructor)]
    pub fn new(buf: Vec<u8>) -> Self {
        set_panic_hook();
        let tagged_file = Probe::new(Cursor::new(buf.clone()))
            .guess_file_type()
            .expect("fail to guess file type")
            .read()
            .expect("fail to read file");

        MetaFile {
            buffer: buf,
            tag: tagged_file.primary_tag().cloned().unwrap_or_else(|| {
                tagged_file
                    .first_tag()
                    .cloned()
                    .expect("ERROR: No tags found!")
            }),
        }
    }

    // other tags...

    #[wasm_bindgen]
    pub fn save(&mut self) -> () {
        // is .clone Avoidable?
        let mut buf = Cursor::new(self.buffer.clone());

        let _ = self.tag.save_to(&mut buf, WriteOptions::default());
        self.buffer = buf.into_inner();
    }
}
```

the suggested solution is below, but errors `the trait bound '&mut Vec<u8>: Truncate' is not satisfied`

```rs
#[wasm_bindgen]
pub fn save(&mut self) -> () {
    set_panic_hook();
    let mut buf = Cursor::new(&mut self.buffer);
    self.tag.save_to(&mut buf, WriteOptions::default());
}
```

What is the correct way to implement it?

**Reply by Serial-ATA** [reply]
That's my bad, I forgot to add `impl<T: Truncate> Truncate for &mut T`. I will get that fixed in 0.19.1 in about an hour.

Just a note, you can avoid the clone when reading the `TaggedFile` as well:

```rs
let tagged_file = Probe::new(Cursor::new(&buf[..]))
    .guess_file_type()
    .expect("fail to guess file type")
    .read()
    .expect("fail to read file");
```

**Reply by Serial-ATA** [reply]
0.19.1 is out now, you should be able to write to a `Cursor<&mut Vec<u8>>` now.

**Reply by jerry7536** [reply]
another error 😂 any suggestion?
```sh
error[E0277]: the trait bound `&mut Vec<u8>: lofty::io::Length` is not satisfied
  --> src\tag.rs:50:18
   |
50 |         self.tag.save_to(&mut buf, WriteOptions::default());
   |                  ^^^^^^^ the trait `lofty::io::Length` is not implemented for `&mut Vec<u8>`, which is required by `std::io::Cursor<&mut Vec<u8>>: lofty::io::Length`
   |
   = help: the trait `lofty::io::Length` is implemented for `Vec<u8>`
   = note: `lofty::io::Length` is implemented for `&Vec<u8>`, but not for `&mut Vec<u8>`
   = note: required for `std::io::Cursor<&mut Vec<u8>>` to implement `lofty::io::Length`
```

**Reply by Serial-ATA** [reply]
That's my fault again. I thought `impl<T: Length> Truncate for &T` would've applied for `&mut T` as well. I'll have to release 0.19.2 soon. This time I'll make a test.

**Reply by Serial-ATA** [reply, marked as answer]
Should be good now, try with 0.19.2

**Reply by jerry7536** [reply]
Yeah, that works!
Thanks for your great work!

**Comment by jerry7536** [comment]
Finally open source it: https://github.com/subframe7536/music-metadata-wasm
